### PR TITLE
fix: skip negative scale checks for creating decimals

### PR DESCRIPTION
**Pull request by kazuyukitanimura**
## Which issue does this PR close?

Part of #679 and #670

## Rationale for this change

This PR improves the getDecimal performance

## What changes are included in this PR?

The new `createDecimal` method skips the negative scale check

## How are these changes tested?

Existing tests

### CODE REVIEW

File: common/src/main/java/org/apache/comet/vector/CometDictionary.java [hunk 1]
```diff
@@ -30,7 +30,7 @@ public class CometDictionary implements AutoCloseable {
   private final int numValues;
 
   /** Decoded dictionary values. We only need to copy values for decimal type. */
-  private ByteArrayWrapper[] binaries;
+  private volatile ByteArrayWrapper[] binaries;
```
Comment (kazuyukitanimura) [inline common/src/main/java/org/apache/comet/vector/CometDictionary.java:33]: This is not directly related but forgot to add based on https://github.com/apache/datafusion-comet/pull/705#discussion_r1688697138

**Comment by kazuyukitanimura** [comment]
SF=100

## getDecimal(Long)
### Before
Screenshot 2024-07-26 at 11 39 30 AM
### After
Screenshot 2024-07-26 at 11 39 51 AM
## getDecimal(128)
### Before
Screenshot 2024-07-26 at 11 40 13 AM
### After
Screenshot 2024-07-26 at 11 40 29 AM

File: common/src/main/java/org/apache/comet/vector/CometVector.java [hunk 2]
```diff
@@ -73,31 +86,35 @@ public boolean isFixedLength() {
...
-                + " and scale: "
-                + scale);
-      }
+      return createDecimal(javaDecimal, precision, scale);
     }
   }
 
+  /** This method skips the negative scale check, otherwise the same as Decimal.createUnsafe(). */
+  private Decimal createDecimal(long unscaled, int precision, int scale) {
+    Decimal dec = new Decimal();
```
Comment (parthchandra) [inline common/src/main/java/org/apache/comet/vector/CometVector.java:102]: Seems the check for negative scale is a feature for ANSI correctness? https://issues.apache.org/jira/browse/SPARK-30252

File: common/src/main/java/org/apache/comet/vector/CometVector.java [hunk 3]
```diff
@@ -73,31 +86,35 @@ public boolean isFixedLength() {
...
+    dec.org$apache$spark$sql$types$Decimal$$longVal_$eq(unscaled);
+    dec.org$apache$spark$sql$types$Decimal$$_precision_$eq(precision);
+    dec.org$apache$spark$sql$types$Decimal$$_scale_$eq(scale);
+    return dec;
+  }
+
+  /** This method skips a few checks, otherwise the same as Decimal.apply(). */
+  private Decimal createDecimal(BigDecimal value, int precision, int scale) {
+    Decimal dec = new Decimal();
+    Platform.putObjectVolatile(dec, decimalValOffset, new scala.math.BigDecimal(value));
```
Comment (parthchandra) [inline common/src/main/java/org/apache/comet/vector/CometVector.java:112]: The spark version has some additional checks for rounding and precision. Do we not need those?
```
    this.decimalVal = decimal.setScale(scale, ROUND_HALF_UP)
    if (decimalVal.precision > precision) {
      throw QueryExecutionErrors.decimalPrecisionExceedsMaxPrecisionError(
        decimalVal.precision, precision)
    }

```

File: common/src/main/java/org/apache/comet/vector/CometVector.java (hunk 2 above)
Comment (kazuyukitanimura) [reply common/src/main/java/org/apache/comet/vector/CometVector.java:102]: It does not need to do the check every read, we are running ANSI tests in spark_sql_test_ansi.yml

File: common/src/main/java/org/apache/comet/vector/CometVector.java (hunk 3 above)
Comment (kazuyukitanimura) [reply common/src/main/java/org/apache/comet/vector/CometVector.java:112]: No we do not need the precision check because it is done in advance

**Comment by kazuyukitanimura** [comment]
@andygrove @comphead @huaxingao @parthchandra @viirya 
Any other feedbacks?

File: common/src/main/java/org/apache/comet/vector/CometVector.java (hunk 3 above)
Comment (parthchandra) [reply common/src/main/java/org/apache/comet/vector/CometVector.java:112]: Just for my knowledge, where?

**Review by parthchandra — APPROVED** [review]
lgtm

**Review by andygrove — APPROVED** [review]
